### DISCUSSION TITLE
One or more objects in request body?

### DISCUSSION THREAD

**Discussion by kevin-hanselman** [fastapi/fastapi#8176] [Questions] [answered]
**Description**

I'd like to have an endpoint in my API that accepts one or more objects in the request body. For example, I would like both of these to be valid request bodies:

``` javascript
{
  "name": "foo"
}
```

``` javascript
[
  {"name": "foo"},
  {"name": "bar"}
]
```

I'm not sure how to do this in fastapi/pydantic. Is it even possible? Any advice?

**Comment by dmontagu** [comment]
I haven't yet seen anything like this implemented in FastAPI, but I think a reasonable starting point would be to try to use the recently-added [custom root model](https://pydantic-docs.helpmanual.io/#custom-root-types) feature in pydantic (I am not sure whether FastAPI is compatible with the appropriate version of pydantic yet). That said, it wouldn't surprise me if some deeper modifications to FastAPI were required to achieve compatibility with this feature.

Note that it would be straightforward to handle either case if contained inside an object wrapper, i.e.:
```
{
    "name_data": {
        "name": "foo"
    }
}
```
vs.
```
{
    "name_data": [
        {"name": "foo"},
        {"name": "bar"}
    ]
}
```
which (I believe) could be implemented via
```python
class Name(BaseModel):
    name: str

class NameData(BaseModel):
    name_data: Union[Name, List[Name]]

**Comment by kevin-hanselman** [comment]
@dmontagu Thanks for the response. Yes, I'm aware of the pattern you mention above, but I want to avoid wrapping my object(s) as such. I've also run into problems with Pydantic when it tries to marshall Unions with collection types like in your example above; I might try running your example to see if it works. Lastly, I've also looked at the custom root models in Pydantic, and I had some troubles getting a custom validator to work with one (and your concern about FastAPI compatibility is valid).

**Comment by dmontagu** [comment]
@kevlar1818 

`Union` is currently somewhat finicky in pydantic due to the way it is implemented -- the order matters because it will try to parse types in the provided order. For what it's worth, if you have issues with pydantic `Union` parsing, I would encourage you to open a github issue over on the pydantic repository -- if the problem _isn't_ a bug (which it might be), you'll probably get a suggestion for how to accomplish your goal.

-----
For what it's worth, the following script does work for me:
```python
from typing import List

from pydantic import BaseModel, Union


class Name(BaseModel):
    name: str


class NameData(BaseModel):
    name_data: Union[Name, List[Name]]


data1 = {
    "name_data": {
        "name": "foo"
    }
}

data2 = {
    "name_data": [
        {"name": "foo"},
        {"name": "bar"}
    ]
}

print(NameData(**data1))
# NameData name_data=<Name name='foo'>

print(NameData(**data2))
# NameData name_data=[<Name name='foo'>, <Name name='bar'>]
```
-----
If your issue with the container approach is that you don't want to write a container model for every list-or-object type you want to have, you can leverage `pydantic.generics.GenericModel`:

```python
from typing import List, TypeVar, Generic

from pydantic import BaseModel, Union
from pydantic.generics import GenericModel

T = TypeVar("T")


class Name(BaseModel):
    name: str


class ListOrObject(GenericModel, Generic[T]):
    contained: Union[T, List[T]]


data1 = {
    "contained": {
        "name": "foo"
    }
}

data2 = {
    "contained": [
        {"name": "foo"},
        {"name": "bar"}
    ]
}

data3 = {
    "contained": "string1"
}

data4 = {
    "contained": ["string1", "string2"]
}

print(ListOrObject[Name](**data1).contained)
# Name name='foo'

print(ListOrObject[Name](**data2).contained)
# [<Name name='foo'>, <Name name='bar'>]

print(ListOrObject[str](**data3).contained)
# string1

print(ListOrObject[str](**data4).contained)
# ['string1', 'string2']
```
(The above script runs for me as is.)

I believe this plays nicely with FastAPI as well.

**Comment by kevin-hanselman** [comment]
Thanks for the detailed response! I'll consider these options and decide how to proceed. I'll mark this issue as closed unless someone else objects.

**Comment by tiangolo** [comment]
Thanks for the help @dmontagu 🌮!

Thanks @kevlar1818 for reporting back and closing the issue.

**Comment by YuriiMotov** [comment, marked as answer]
The following works just fine in current version (FastAPI 0.116.0, Pydantic 1.10.22 / 2.11.7):
```python
from fastapi import FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    name: str


@app.post("/")
async def add_items(item_or_items: Item | list[Item]):
    return item_or_items


client = TestClient(app)


def test_add_single_item():
    response = client.post("/", json={"name": "apple"})
    assert response.status_code == 200
    assert response.json() == {"name": "apple"}


def test_add_multiple_items():
    items = [{"name": "apple"}, {"name": "banana"}]
    response = client.post("/", json=items)
    assert response.status_code == 200
    assert response.json() == items


def test_add_invalid_item():
    response = client.post("/", json={"not_name": "apple"})
    assert response.status_code == 422


def test_add_empty_list():
    response = client.post("/", json=[])
    assert response.status_code == 200
    assert response.json() == []

```
